include/sparrow/array/external_array_data.hpp

```diff
@@ -0,0 +1,370 @@
+
+        void build_children();
+        void build_dictionary();
+
+        impl::external_wrapper<ArrowSchema> m_schema;
+        impl::external_wrapper<ArrowArray> m_array;
+
+        std::vector<external_array_data> m_children;
+        value_ptr<external_array_data> m_dictionary;
+    };
+
+    data_descriptor type_descriptor(const external_array_data& data);
+    external_array_data::length_type length(const external_array_data& data);
+    std::int64_t offset(const external_array_data& data);
+
+    external_array_data::bitmap_type bitmap(const external_array_data& data);
+
+    std::size_t buffers_size(const external_array_data& data);
+    external_array_data::buffer_type buffer_at(const external_array_data& data, std::size_t i);
+
+    std::size_t child_data_size(const external_array_data& data);
+    const external_array_data& child_data_at(const external_array_data& data, std::size_t i);
+
+    const value_ptr<external_array_data>& dictionary(const external_array_data& data);
+
+
+    // `external_array_data` must always be usable as a data-storage that layout implementations can use.
+    static_assert(data_storage<external_array_data>);
+
+
+    /***********************************
+     * external_wrapper implementation *
+     ***********************************/
+
+    namespace impl
+    {
+        template <any_arrow_c_interface T>
+        template <class U>
+            requires std::convertible_to<U, T>
+        external_wrapper<T>::external_wrapper(U&& u, ownership ownership_model)
+        {
+            if (ownership_model == ownership::owning)
+            {
+                p_data = std::shared_ptr<T>(new T(std::forward<U>(u)), full_ownership_deleter{});
+            }
+            else
+            {
+                p_data = std::shared_ptr<T>(new T(std::forward<U>(u))); // default shared_ptr deleter
+            }
+        }
+
+        template <any_arrow_c_interface T>
+        template <class U>
+            requires std::is_pointer_v<std::remove_reference_t<U>> and std::convertible_to<U, T*>
+        external_wrapper<T>::external_wrapper(U ptr, ownership ownership_model)
+        {
+            if (ownership_model == ownership::owning)
+            {
+                p_data = std::shared_ptr<T>(ptr, only_release_arrow_data{});
+            }
+            else
+            {
+                p_data = std::shared_ptr<T>(ptr, dont_release_anything{});
+            }
+        }
+
+        template <any_arrow_c_interface T>
+        T& external_wrapper<T>::data()
+        {
+            return *p_data;
+        }
+
+        template <any_arrow_c_interface T>
+        const T& external_wrapper<T>::data() const
+        {
+            return *p_data;
+        }
+
+        template <any_arrow_c_interface T>
+        void external_wrapper<T>::only_release_arrow_data::operator()(T* t) const
+        {
+            if (t->release)
+            {
+                t->release(t);
+                t->release = nullptr;
```

Comment:
Can be removed as the call of the function set the t->release function pointer to nullptr